cdap-watchdog/src/main/java/co/cask/cdap/logging/save/CheckpointManager.java

```diff
@@ -37,50 +41,84 @@
+                                                                                (TransactionAware) table);
+        return txExecutor.execute(func, table);
+      } else {
+        throw new RuntimeException(String.format("Table %s is not TransactionAware, " +
+                                                   "Exception while trying to cast it to TransactionAware. " +
+                                                   "Please check why the table is not TransactionAware", table));
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Error accessing %s table", tableUtil.getMetaTableName()), e);
+    }
+  }
+
+  public void saveCheckpoint(final Map<Integer, Checkpoint> checkpointMap) throws Exception {
+    LOG.trace("Saving checkpoints for partitions {}", checkpointMap);
+
+    // if the checkpoints have not changed, we skip writing to table and return.
+    if (lastCheckpoint.equals(checkpointMap)) {
+      return;
+    }
+
+    execute(new TransactionExecutor.Function<Table, Void>() {
       @Override
-      public DatasetContext<Table> get() {
-        try {
-          return DatasetContext.of(tableUtil.getMetaTable());
-        } catch (Exception e) {
-          // there's nothing much we can do here
-          throw Throwables.propagate(e);
+      public Void apply(Table table) throws Exception {
+        for (Map.Entry<Integer, Checkpoint> entry : checkpointMap.entrySet()) {
+          byte[] key = Bytes.add(rowKeyPrefix, Bytes.toBytes(entry.getKey()));
+          Checkpoint checkpoint = entry.getValue();
+          table.put(key, OFFSET_COLNAME, Bytes.toBytes(checkpoint.getNextOffset()));
+          table.put(key, MAX_TIME_COLNAME, Bytes.toBytes(checkpoint.getMaxEventTime()));
         }
+        // update last checkpoint
+        lastCheckpoint = ImmutableMap.copyOf(checkpointMap);
+        return null;
       }
     });
   }
 
-
-  public void saveCheckpoint(final int partition, final Checkpoint checkpoint) throws Exception {
-    LOG.trace("Saving checkpoint {} for partition {}", checkpoint, partition);
-    mds.execute(new TransactionExecutor.Function<DatasetContext<Table>, Void>() {
+  public Map<Integer, Checkpoint> getCheckpoint(final Set<Integer> partitions) throws Exception {
+    return execute(new TransactionExecutor.Function<Table, Map<Integer, Checkpoint>>() {
       @Override
-      public Void apply(DatasetContext<Table> ctx) throws Exception {
-        Table table = ctx.get();
-        byte[] key = Bytes.add(rowKeyPrefix, Bytes.toBytes(partition));
-        table.put(key, OFFSET_COLNAME, Bytes.toBytes(checkpoint.getNextOffset()));
-        table.put(key, MAX_TIME_COLNAME, Bytes.toBytes(checkpoint.getMaxEventTime()));
-        return null;
+      public Map<Integer, Checkpoint> apply(Table table) throws Exception {
+        Map<Integer, Checkpoint> checkpointMap1 = new HashMap<Integer, Checkpoint>();
```

Comment:
Also, throughout, can we remove ``Map`` from the var names? The type already indicates that it is a map.